### PR TITLE
Update our syntax list types to be "collection builder" types.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/71596

API change suggested and approved in https://github.com/dotnet/roslyn/issues/71596.

This makes

```
SyntaxList<T>
SeparatedSyntaxList<T>
SyntaxTriviaList
SyntaxTokenList
SyntaxNodeOrTokenList
```

all creatable using collection expressions.

### CODE REVIEW

File: src/Compilers/Test/Core/CollectionBuilderAttribute.cs [hunk 1]
```diff
@@ -2,6 +2,8 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#if false
```
Comment (sharwell) [inline src/Compilers/Test/Core/CollectionBuilderAttribute.cs:5]: 😕 Why was this needed?

Comment (CyrusNajmabadi) [reply src/Compilers/Test/Core/CollectionBuilderAttribute.cs:5]: The type was added at the core layer.  So this one became unneeded

File: src/Compilers/CSharp/Test/Syntax/Syntax/SeparatedSyntaxListTests.cs [hunk 2]
```diff
@@ -44,11 +44,15 @@ public void EnumeratorEquality()
...
         }
 
-        [WorkItem(308077, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/308077")]
-        [Fact]
+        [Fact, WorkItem("http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/308077")]
         public void TestSeparatedListInsert()
         {
-            var list = SyntaxFactory.SeparatedList<ExpressionSyntax>();
+            TestSeparatedListInsertWorker(SyntaxFactory.SeparatedList<ExpressionSyntax>());
+            TestSeparatedListInsertWorker([]);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SeparatedSyntaxListTests.cs:51]: to add tests, i went to all the SyntaxFactory methods that create the different list types and i went and found where they were used in tests.  i then tried to update at least two tests.  A test that called the empty factory method, and tests that called the singleton or multi item factory methods.

File: src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryLambdaExpression/CSharpRemoveUnnecessaryLambdaExpressionDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -293,7 +293,7 @@ private static SeparatedSyntaxList<ParameterSyntax> GetParameters(AnonymousFunct
             => expression switch
             {
                 AnonymousMethodExpressionSyntax anonymousMethod => anonymousMethod.ParameterList?.Parameters ?? default,
-                SimpleLambdaExpressionSyntax simpleLambda => SyntaxFactory.SingletonSeparatedList(simpleLambda.Parameter),
+                SimpleLambdaExpressionSyntax simpleLambda => [simpleLambda.Parameter],
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryLambdaExpression/CSharpRemoveUnnecessaryLambdaExpressionDiagnosticAnalyzer.cs:296]: Updated IDE to get lots of test coverage of the new collection expr support.

File: src/Compilers/Core/Portable/Syntax/GreenNode.cs [hunk 4]
```diff
@@ -846,15 +846,15 @@ private static bool EquivalentToInternal(GreenNode node1, GreenNode node2)
                 case 1:
                     return select(list[0]);
                 case 2:
-                    return SyntaxList.List(select(list[0]), select(list[1]));
+                    return Syntax.InternalSyntax.SyntaxList.List(select(list[0]), select(list[1]));
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/GreenNode.cs:849]: had to qualify since there is now a public type called SyntaxList (which contains the factory method).

File: src/Compilers/Core/Portable/Syntax/SyntaxNodeOrTokenList.cs [hunk 5]
```diff
@@ -63,6 +65,26 @@ public SyntaxNodeOrTokenList(params SyntaxNodeOrToken[] nodesAndTokens)
...
+
+            return new SyntaxNodeOrTokenList(CreateNode(nodesAndTokens));
+        }
+
+        private static SyntaxNode? CreateNode(ReadOnlySpan<SyntaxNodeOrToken> nodesAndTokens)
+        {
+            if (nodesAndTokens == null)
+            {
+                throw new ArgumentNullException(nameof(nodesAndTokens));
+            }
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/SyntaxNodeOrTokenList.cs:81]:
```suggestion
            if (nodesAndTokens == null)
                throw new ArgumentNullException(nameof(nodesAndTokens));
```

File: src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs [hunk 6]
```diff
@@ -8,11 +8,39 @@
...
 using System.Linq;
+using System.Runtime.CompilerServices;
 using Microsoft.CodeAnalysis.Text;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis
 {
+    public static class SeparatedSyntaxList
+    {
+        public static SeparatedSyntaxList<TNode> Create<TNode>(ReadOnlySpan<TNode> nodes) where TNode : SyntaxNode
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs:19]: i kept the parameter name the same as SyntaxFactory.SeparatedList.

File: src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs [hunk 7]
```diff
@@ -8,15 +8,36 @@
...
+using System.Runtime.CompilerServices;
 using Microsoft.CodeAnalysis.Syntax;
 using Microsoft.CodeAnalysis.Text;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis
 {
+    public static class SyntaxList
+    {
+        public static SyntaxList<TNode> Create<TNode>(ReadOnlySpan<TNode> nodes) where TNode : SyntaxNode
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs:20]: i kept the parameter name the same as SyntaxFactory.List.

File: src/Compilers/Core/Portable/Syntax/SyntaxNodeOrTokenList.cs [hunk 8]
```diff
@@ -63,6 +65,26 @@ public SyntaxNodeOrTokenList(params SyntaxNodeOrToken[] nodesAndTokens)
         {
         }
 
+        public static SyntaxNodeOrTokenList Create(ReadOnlySpan<SyntaxNodeOrToken> nodesAndTokens)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/SyntaxNodeOrTokenList.cs:68]: i kept the parameter name the same as SyntaxFactory.NodeOrTokenList.

File: src/Compilers/Core/Portable/Syntax/SyntaxTokenList.cs [hunk 9]
```diff
@@ -59,7 +61,21 @@ public SyntaxTokenList(IEnumerable<SyntaxToken> tokens)
         {
         }
 
+        public static SyntaxTokenList Create(ReadOnlySpan<SyntaxToken> tokens)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/SyntaxTokenList.cs:64]: i kept the parameter name the same as SyntaxFactory.TokenList.

File: src/Compilers/Core/Portable/Syntax/SyntaxTriviaList.cs [hunk 10]
```diff
@@ -65,7 +67,21 @@ public SyntaxTriviaList(IEnumerable<SyntaxTrivia>? trivias)
         {
         }
 
+        public static SyntaxTriviaList Create(ReadOnlySpan<SyntaxTrivia> trivias)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/SyntaxTriviaList.cs:70]: i kept the parameter name the same as SyntaxFactory.TriviaList.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler @dotnet/roslyn-ide  this is ready for review.

**Review by AlekseyTs — CHANGES_REQUESTED** [review]
Please revert changes under Compilers with exception of the new APIs. There could be other exceptions, but at the moment they are getting "lost" among style changes, which are not welcome.

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs [hunk 11]
```diff
@@ -131,7 +131,7 @@ private void AnalyzeArrayInitializerExpression(SyntaxNodeAnalysisContext context
         // Have to actually examine what would happen when we do the replacement, as the replaced value may interact
         // with inference based on the values within.
         var replacementCollectionExpression = CollectionExpression(
-            SeparatedList<CollectionElementSyntax>(initializer.Expressions.Select(ExpressionElement)));
+            [.. initializer.Expressions.Select(ExpressionElement)]);
```
Comment (RikkiGibson) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs:134]: I didn't expect to see a space between `..` and `initializer` here. Is that included intentionally by the formatter?

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs:134]: Yes. Very intentional. Matches what we do with list patterns. Also how we doc this everywhere.  It's not an operator that binds to the thing on the right.  It affects the entire expression, so it needs a space.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler ptal.

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNodeTests.cs [hunk 12]
```diff
@@ -679,7 +679,9 @@ public void TestFindTokenInLargeList()
                 missingArgument, missingComma,
                 argument);
 
-            var argumentList = SyntaxFactory.ArgumentList(SyntaxFactory.SeparatedList<ArgumentSyntax>(SyntaxFactory.NodeOrTokenList(nodesAndTokens)));
+            var argumentList = SyntaxFactory.ArgumentList(SyntaxFactory.SeparatedList<ArgumentSyntax>(collectionExpression
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNodeTests.cs:682]: >SyntaxFactory.SeparatedList<ArgumentSyntax>( [](http://example.com/codeflow?start=58&length=44)

nit: Do we need the `SyntaxFactory.SeparatedList<ArgumentSyntax>(` part in the collection expression case?

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNodeTests.cs:682]: probably not.  but i was trying to test out the NodeOrTokenList bypass.

File: src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs [hunk 13]
```diff
@@ -8,11 +8,39 @@
...
+
+            var builder = new CodeAnalysis.Syntax.SeparatedSyntaxListBuilder<TNode>(nodes.Length);
+
+            builder.Add(nodes[0]);
+
+            var separator = nodes[0].Green.CreateSeparator(nodes[0]);
+
+            for (int i = 1, n = nodes.Length; i < n; i++)
+            {
+                builder.AddSeparator(separator);
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs:35]: How did you decide to add separators vs. just inserting the provided nodes?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 61). Only reviewed the changes under the Compiler folder

File: src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs (hunk 13 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs:35]: this is the overload that takes in nodes-only, but is producing a separated list.  the semantics are that we must separate the nodes with the language separator.  Note: this is how teh existing SyntaxFactory.SeparatedList helper worked (though it could hardcode in directly the comma token at the lang specific level).  At this level, we need the virtual call to get teh separator as it will be different from vb to c#.

File: src/Compilers/Core/Portable/Syntax/SyntaxTriviaList.cs [hunk 14]
```diff
@@ -65,7 +67,21 @@ public SyntaxTriviaList(IEnumerable<SyntaxTrivia>? trivias)
...
+        {
+            if (trivias.Length == 0)
+                return default;
+
+            if (trivias.Length == 1)
+                return new SyntaxTriviaList(trivias[0]);
+
+            return new SyntaxTriviaList(token: default, CreateNode(trivias), position: 0, index: 0);
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/Syntax/SyntaxTriviaList.cs:78]: It looks like these impls are imitating the behavior you would have gotten for 0/1/n arguments to similar SyntaxFactory methods? Is that accurate?

File: src/Compilers/Core/Portable/Syntax/SyntaxTriviaList.cs [hunk 15]
```diff
@@ -65,7 +67,21 @@ public SyntaxTriviaList(IEnumerable<SyntaxTrivia>? trivias)
...
+        }
+
         private static GreenNode? CreateNode(SyntaxTrivia[]? trivias)
+            => CreateNode(trivias.AsSpan());
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/Syntax/SyntaxTriviaList.cs:82]: nit: seems like this is going from Array to Span to ReadOnlySpan, when we could just remove the call to AsSpan and use the implicit conversion straight from Array to ReadOnlySpan.

File: src/Compilers/Core/Portable/Syntax/CollectionBuilderAttribute.cs [hunk 16]
```diff
@@ -7,7 +7,7 @@
 namespace System.Runtime.CompilerServices
 {
     [AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Interface)]
-    public sealed class CollectionBuilderAttribute : Attribute
+    internal sealed class CollectionBuilderAttribute : Attribute
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/Syntax/CollectionBuilderAttribute.cs:10]: Prior to this PR, was it accidental that this was declared public? Or has something changed so that only now we want it to be internal?

**Review by RikkiGibson — APPROVED** [review]
Changes under Compilers LGTM aside from some minor questions/nits

File: src/Compilers/Core/Portable/Syntax/SyntaxTriviaList.cs (hunk 14 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Syntax/SyntaxTriviaList.cs:78]: correct.

File: src/Compilers/Core/Portable/Syntax/SyntaxTriviaList.cs (hunk 15 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Syntax/SyntaxTriviaList.cs:82]: oh.  let me change that.

File: src/Compilers/Core/Portable/Syntax/CollectionBuilderAttribute.cs (hunk 16 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Syntax/CollectionBuilderAttribute.cs:10]: prior to this pr this was only defined in the test layer.  now it's pulled down into the actual compiler layer effectively as a polyfill.  so we're saying:

prior to .net8, we add our own attribute here, so that our compiler can see this when people target this api.
after .net8, we pick up the actual real attribute.
